In [1]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
train = pd.read_csv('ML_train.csv')
test = pd.read_csv('ML_test.csv')

In [3]:
train.head()

Unnamed: 0                                               Text  Sentiment
0           0  movie is funny suitable age is definitely fami...          6
1           1  old commercial blank audio cassette tag line w...          7
2           2  cinemascope color cinematography leon shamroy ...          4
3           3  get film possible will find really good perfor...          6
4           4  soundtrack is bit dated story is relevant yous...          6

In [4]:
train = train.drop(columns=['Unnamed: 0'])
train.head()

Text  Sentiment
0  movie is funny suitable age is definitely fami...          6
1  old commercial blank audio cassette tag line w...          7
2  cinemascope color cinematography leon shamroy ...          4
3  get film possible will find really good perfor...          6
4  soundtrack is bit dated story is relevant yous...          6

In [5]:
trial_train = train.copy()
trial_train.head()

Text  Sentiment
0  movie is funny suitable age is definitely fami...          6
1  old commercial blank audio cassette tag line w...          7
2  cinemascope color cinematography leon shamroy ...          4
3  get film possible will find really good perfor...          6
4  soundtrack is bit dated story is relevant yous...          6

In [6]:
def good_bad(row):
  if row > 4:
    return 1
  else:
    return 0

In [7]:
trial_train['Good'] = trial_train['Sentiment'].apply(good_bad)
trial_train.head()

Text  Sentiment  Good
0  movie is funny suitable age is definitely fami...          6     1
1  old commercial blank audio cassette tag line w...          7     1
2  cinemascope color cinematography leon shamroy ...          4     0
3  get film possible will find really good perfor...          6     1
4  soundtrack is bit dated story is relevant yous...          6     1

In [8]:
class2_X_train = trial_train['Text']
class2_y_train = trial_train['Good']

In [9]:
trial_test = test.copy()
trial_test.head()

Unnamed: 0                                               Text  Sentiment
0           0  would highly recommend seeing movie viewing wi...          1
1           1  see everyone love film much true doe have good...          1
2           2  damn thought would seen bad western cannot top...          1
3           3  well certainly stunned believe someone made an...          1
4           4  jefferey dahmer wa one sick guy is much say al...          1

In [10]:
trial_test['Good'] = trial_test['Sentiment'].apply(good_bad)
trial_test.head()

Unnamed: 0                                               Text  Sentiment  \
0           0  would highly recommend seeing movie viewing wi...          1   
1           1  see everyone love film much true doe have good...          1   
2           2  damn thought would seen bad western cannot top...          1   
3           3  well certainly stunned believe someone made an...          1   
4           4  jefferey dahmer wa one sick guy is much say al...          1   

   Good  
0     0  
1     0  
2     0  
3     0  
4     0

In [11]:
class2_X_test = trial_test['Text']
class2_y_test = trial_test['Good']

# Vectorization

BOW Vectorizer

In [12]:
bow_vectorizer = CountVectorizer()
bow_vectorizer.fit(class2_X_train)

class2_bow_X_train = bow_vectorizer.transform(class2_X_train)
class2_bow_X_test = bow_vectorizer.transform(class2_X_test)

TF-IDF

In [13]:

tfidf_vectorizer = TfidfVectorizer()
tfidf_vectorizer.fit(class2_X_train)

class2_tfidf_X_train = tfidf_vectorizer.transform(class2_X_train)
class2_tfidf_X_test = tfidf_vectorizer.transform(class2_X_test)

# Tuning Naive Bayes with BOW

In [15]:
param_grid = {
    'alpha': [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
    'force_alpha': [True, False],
    'fit_prior': [True, False]
}

In [16]:
nb_model = MultinomialNB()
clf = GridSearchCV(nb_model, param_grid=param_grid, cv=3, verbose=4, scoring='accuracy')

In [17]:
best_clf = clf.fit(class2_tfidf_X_train, class2_y_train)

Fitting 3 folds for each of 44 candidates, totalling 132 fits
[CV 1/3] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.743 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/3] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.750 total time=   0.1s
[CV 3/3] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.980 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/3] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.857 total time=   0.1s
[CV 2/3] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.861 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 3/3] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.980 total time=   0.1s
[CV 1/3] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.743 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/3] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.750 total time=   0.1s
[CV 3/3] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.980 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/3] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.857 total time=   0.1s
[CV 2/3] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.861 total time=   0.1s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 3/3] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.980 total time=   0.1s
[CV 1/3] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.903 total time=   0.1s
[CV 2/3] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.903 total time=   0.1s
[CV 3/3] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.944 total time=   0.1s
[CV 1/3] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.903 total time=   0.1s
[CV 2/3] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.903 total time=   0.1s
[CV 3/3] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.944 total time=   0.1s
[CV 1/3] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.903 total time=   0.1s
[CV 2/3] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.903 total time=   0.1s
[CV 3/3] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.944 total time=   0.1s
[CV 1/3] END alpha=0.1, fit_prior=False, force_alpha=False;, score=0.903 total time=   0.1s


In [20]:
# best parameters
best_clf.best_params_

{'alpha': 0.1, 'fit_prior': True, 'force_alpha': True}

In [21]:
# highest accuracy score
best_clf.best_score_

0.916584757317664

Function to help us evaluate model performance

In [14]:
def train_and_eval(model, trainX, trainY, testX, testY):

    # training the model
    fitted_model = model.fit(trainX, trainY)

    # getting predictions
    y_preds_train = fitted_model.predict(trainX)
    y_preds_test = fitted_model.predict(testX)

    # evaluating the model
    print()
    print(model)
    print(f"Train accuracy score : {accuracy_score(trainY, y_preds_train)}")
    print(f"Test accuracy score : {accuracy_score(testY, y_preds_test)}")
    print(classification_report(testY, y_preds_test))
    print('\n',40*'-')

In [23]:
best_nb = MultinomialNB(alpha = 0.1, fit_prior = True, force_alpha = True)

In [26]:
train_and_eval(best_nb, class2_bow_X_train, class2_y_train, class2_bow_X_test, class2_y_test)


MultinomialNB(alpha=0.1, force_alpha=True)
Train accuracy score : 0.9499567794517165
Test accuracy score : 0.8514081027667985
              precision    recall  f1-score   support

           0       0.84      0.87      0.85      8096
           1       0.86      0.84      0.85      8096

    accuracy                           0.85     16192
   macro avg       0.85      0.85      0.85     16192
weighted avg       0.85      0.85      0.85     16192


 ----------------------------------------


Best Naive Bayes model with BOW gave us 85.1% accuracy on test data

# Tuning Naive Bayes with TF-IDF

In [15]:
param_grid = {
    'alpha': [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
    'force_alpha': [True, False],
    'fit_prior': [True, False]
}

In [16]:
nb_model = MultinomialNB()
clf = GridSearchCV(nb_model, param_grid=param_grid, cv=3, verbose=4, scoring='accuracy')

In [17]:
best_clf = clf.fit(class2_tfidf_X_train, class2_y_train)

Fitting 3 folds for each of 44 candidates, totalling 132 fits


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 1/3] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.743 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/3] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.750 total time=   0.2s
[CV 3/3] END alpha=0.0, fit_prior=True, force_alpha=True;, score=0.980 total time=   0.2s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/3] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.857 total time=   0.2s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/3] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.861 total time=   0.3s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 3/3] END alpha=0.0, fit_prior=True, force_alpha=False;, score=0.980 total time=   0.4s
[CV 1/3] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.743 total time=   0.2s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(


[CV 2/3] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.750 total time=   0.2s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:907: RuntimeWarning: divide by zero encountered in log
  self.feature_log_prob_ = np.log(smoothed_fc) - np.log(
/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 3/3] END alpha=0.0, fit_prior=False, force_alpha=True;, score=0.980 total time=   0.2s
[CV 1/3] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.857 total time=   0.2s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 2/3] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.861 total time=   0.2s
[CV 3/3] END alpha=0.0, fit_prior=False, force_alpha=False;, score=0.980 total time=   0.2s


/usr/local/lib/python3.10/dist-packages/sklearn/naive_bayes.py:635: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10. Use `force_alpha=True` to keep alpha unchanged.
  warnings.warn(


[CV 1/3] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.903 total time=   0.2s
[CV 2/3] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.903 total time=   0.2s
[CV 3/3] END alpha=0.1, fit_prior=True, force_alpha=True;, score=0.944 total time=   0.1s
[CV 1/3] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.903 total time=   0.1s
[CV 2/3] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.903 total time=   0.2s
[CV 3/3] END alpha=0.1, fit_prior=True, force_alpha=False;, score=0.944 total time=   0.3s
[CV 1/3] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.903 total time=   0.2s
[CV 2/3] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.903 total time=   0.2s
[CV 3/3] END alpha=0.1, fit_prior=False, force_alpha=True;, score=0.944 total time=   0.3s
[CV 1/3] END alpha=0.1, fit_prior=False, force_alpha=False;, score=0.903 total time=   0.2s
[CV 2/3] END alpha=0.1, fit_prior=False, force_alpha=False;, score=0.903 total time=   0.3s


In [18]:
# best parameters
best_clf.best_params_

{'alpha': 0.1, 'fit_prior': True, 'force_alpha': True}

In [19]:
# highest accuracy score
best_clf.best_score_

0.916584757317664

In [20]:
best_nb = MultinomialNB(alpha = 0.1, fit_prior = True, force_alpha = True)

In [24]:
train_and_eval(best_nb, class2_tfidf_X_train, class2_y_train, class2_tfidf_X_test, class2_y_test)


MultinomialNB(alpha=0.1, force_alpha=True)
Train accuracy score : 0.9629846875771796
Test accuracy score : 0.8545578063241107
              precision    recall  f1-score   support

           0       0.85      0.86      0.85      8096
           1       0.86      0.85      0.85      8096

    accuracy                           0.85     16192
   macro avg       0.85      0.85      0.85     16192
weighted avg       0.85      0.85      0.85     16192


 ----------------------------------------


Best Multinomial Naive Bayes with TF-IDF gives us 85.4% accuracy on test data